### PR TITLE
test: add e2e test for updating adapters

### PR DESCRIPTION
**Pull request by bangqipropel**
**Reason for Change**:
Add e2e tests for updating adapters and validation of it
Add controller revision validation to e2e test

**Requirements**

- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

File: test/e2e/preset_test.go [hunk 1]
```diff
@@ -370,6 +398,52 @@ func validateInferenceResource(workspaceObj *kaitov1alpha1.Workspace, expectedRe
 	})
 }
 
+// Logic to validate he annotations of the workspace and deployment, as well as the corresponding controller revision
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:401]: K8s convention for the comment about the method - the sentence starts with the method name.  There is a typo in the comment, "he" -> "the".

File: test/e2e/preset_test.go [hunk 2]
```diff
@@ -246,6 +257,23 @@ func createAndValidateWorkspace(workspaceObj *kaitov1alpha1.Workspace) {
 	})
 }
 
+func updateAndValidateWorkspace(workspaceObj *kaitov1alpha1.Workspace) {
+	By("Creating workspace", func() {
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:261]: Should be "Updating workspace"?

File: test/e2e/preset_test.go [hunk 3]
```diff
@@ -246,6 +257,23 @@ func createAndValidateWorkspace(workspaceObj *kaitov1alpha1.Workspace) {
...
+		Eventually(func() error {
+			return utils.TestingCluster.KubeClient.Update(ctx, workspaceObj)
+		}, utils.PollTimeout, utils.PollInterval).
+			Should(Succeed(), "Failed to update workspace %s", workspaceObj.Name)
+
+		By("Validating workspace creation", func() {
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:267]: Should be "validating workspace update"?

File: test/e2e/preset_test.go [hunk 4]
```diff
@@ -86,6 +89,14 @@ func createCustomWorkspaceWithAdapter(numOfNode int) *kaitov1alpha1.Workspace {
 	return workspaceObj
 }
 
+func updateCustomWorkspaceWithAdapter(workspaceObj *kaitov1alpha1.Workspace, validAdapters []kaitov1alpha1.AdapterSpec) *kaitov1alpha1.Workspace {
+	By("Updating a workspace with adapter", func() {
+		workspaceObj.Inference.Adapters = validAdapters
+		updateAndValidateWorkspace(workspaceObj)
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:95]: updateAndValidateWorkspace is just called once so I think the entire updateAndValidateWorkspace function can be "inline" here.

File: test/e2e/preset_test.go [hunk 5]
```diff
@@ -370,6 +398,52 @@ func validateInferenceResource(workspaceObj *kaitov1alpha1.Workspace, expectedRe
 	})
 }
 
+// Logic to validate he annotations of the workspace and deployment, as well as the corresponding controller revision
+func validateRevision(workspaceObj *kaitov1alpha1.Workspace, revisionStr string) {
+	By("Checking the revisions of the resources", func() {
+		Eventually(func() bool {
+			dep := &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
```
Comment (Fei-Guo) [inline test/e2e/preset_test.go:406]: I don't think you need to add ObjectMeta for the variable that will be used in the Get call. The same for line 421.

File: test/e2e/inference_with_adapters_test.go [hunk 6]
```diff
@@ -104,6 +123,39 @@ func validateAdapterAdded(workspaceObj *kaitov1alpha1.Workspace, deploymentName
...
 }
 
+func validateNoInitContainers(workspaceObj *kaitov1alpha1.Workspace, expectedInitContainers []corev1.Container) {
+	By("Checking the Adapters should be deleted", func() {
+		Eventually(func() bool {
+			var err error
+			var initContainers []corev1.Container
+
+			dep := &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
```
Comment (Fei-Guo) [inline test/e2e/inference_with_adapters_test.go:133]: You don't need to add ObjectMeta for the GET call.

File: test/e2e/inference_with_adapters_test.go [hunk 7]
```diff
@@ -104,6 +123,39 @@ func validateAdapterAdded(workspaceObj *kaitov1alpha1.Workspace, deploymentName
 	})
 }
 
+func validateNoInitContainers(workspaceObj *kaitov1alpha1.Workspace, expectedInitContainers []corev1.Container) {
```
Comment (Fei-Guo) [inline test/e2e/inference_with_adapters_test.go:126]: Why is the method named "NoInitContainers"?

Comment (bangqipropel) [reply test/e2e/inference_with_adapters_test.go:126]: To validate the previous InitContainers has been deleted

File: test/e2e/preset_test.go (hunk 1 above)
Comment (bangqipropel) [reply test/e2e/preset_test.go]: changed all the comment in this file

File: test/e2e/preset_test.go [hunk 8]
```diff
@@ -364,13 +389,48 @@ func validateInferenceResource(workspaceObj *kaitov1alpha1.Workspace, expectedRe
...
+			if err != nil {
+				GinkgoWriter.Printf("Error fetching resource: %v\n", err)
+				return false
+			}
+			workspaceObjHash := workspaceObj.Annotations[WorkspaceHashAnnotation]
+			revision := &appsv1.ControllerRevision{}
+			err = utils.TestingCluster.KubeClient.Get(ctx, client.ObjectKey{
+				Namespace: workspaceObj.Namespace,
+				Name:      fmt.Sprintf("%s-%s", workspaceObj.Name, workspaceObjHash[:5]),
+			}, revision)
```
Comment (ishaansehgal99) [inline test/e2e/preset_test.go:415]: 5 could be a constant

File: test/e2e/inference_with_adapters_test.go [hunk 9]
```diff
@@ -104,6 +123,34 @@ func validateAdapterAdded(workspaceObj *kaitov1alpha1.Workspace, deploymentName
...
```
Comment (ishaansehgal99) [inline test/e2e/inference_with_adapters_test.go:126]: Init containers don't get deleted per my understanding - they get marked completed. Could rename to `validateInitContainersCompleted`?

File: test/e2e/preset_test.go [hunk 10]
```diff
@@ -33,6 +33,9 @@ const (
 	PresetMistral7BInstructModel = "mistral-7b-instruct"
 	PresetPhi2Model              = "phi-2"
 	PresetPhi3Mini128kModel      = "phi-3-mini-128k-instruct"
+	WorkspaceHashAnnotation      = "workspace.kaito.io/hash"
+	// WorkspaceRevisionAnnotation is the Annotations for revision number
```
Comment (ishaansehgal99) [inline test/e2e/preset_test.go:37]: This is a hash right? Or is it a revision number (e.g. "1")? 

I think its a hash be
```suggestion
	// WorkspaceRevisionAnnotation represents a unique hash value of the workspace configuration to track changes
```

**Review by ishaansehgal99 — APPROVED** [review]

File: test/e2e/inference_with_adapters_test.go (hunk 9 above)
Comment (ishaansehgal99) [reply test/e2e/inference_with_adapters_test.go]: nvm resolved offline that new pod is created - as long as we check new deployment has the correct # of init containers, we can remove this

File: docker/adapters/README.md [hunk 11]
```diff
@@ -2,8 +2,8 @@
 
 ## Overview
 
-These files are part of a set used for conducting end-to-end (E2E) testing of an adapter component. The Dockerfile builds an image incorporating the configuration and model files, which is then used within an Init Container for testing. The adapter is training from [dolly-15k-oai-style](https://huggingface.co/datasets/philschmid/dolly-15k-oai-style) dataset
-and was trained using default [qlora-params.yaml](../../charts/kaito/workspace/templates/qlora-params.yaml) 
+These files are part of a set used for conducting end-to-end (E2E) testing of an adapter component. The Dockerfile builds an image incorporating the configuration and model files, which is then used within an Init Container for testing. The adapter1 is training from [dolly-15k-oai-style](https://huggingface.co/datasets/philschmid/dolly-15k-oai-style) dataset
```
Comment (ishaansehgal99) [inline docker/adapters/README.md:5]: Add comment for how adapter2 is made?

**Review by Fei-Guo — APPROVED** [review]

File: docker/adapters/README.md (hunk 11 above)
Comment (bangqipropel) [reply docker/adapters/README.md:5]: They don't have it

**Pull request merged**